File: JavaScript_/Questions/reverseLinkedList.js
//     Given the head of a singly linked list, reverse the list, and return the reversed list.
//     Input: head = [1,2,3,4,5]
//     Output: [5,4,3,2,1]
//     https://leetcode.com/problems/reverse-linked-list/

class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
  }

  append(value) {
    var new_node = new Node(value);

    if (this.head === null) {
      this.head = new_node;
      return;
    }

    var cur_node = this.head;
    while (cur_node.next) {
      cur_node = cur_node.next;
    }

    cur_node.next = new_node;
  }

  printList() {
    var cur_node = this.head;
    while (cur_node) {
      console.log(cur_node.value);
      cur_node = cur_node.next;
    }
  }

  reverse() {
    var prev = null;
    var cur_node = this.head;
    var nxt = null;

    while (cur_node) {
      nxt = cur_node.next;
      cur_node.next = prev;

      prev = cur_node;
      cur_node = nxt;
    }
    this.head = prev;
  }
}

var llist = new LinkedList();
llist.append(1);
llist.append(2);
llist.append(3);
llist.append(4);
llist.append(5);
llist.reverse();
llist.printList();

File: Python_/Questions/reverseLinkedList.py
"""
    Given the head of a singly linked list, reverse the list, and return the reversed list.
    
    Input: head = [1,2,3,4,5]
    Output: [5,4,3,2,1]

    https://leetcode.com/problems/reverse-linked-list/
"""



class Node:
    def __init__(self, value):
        self.value = value
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def append(self, value):
        new_node = Node(value)

        if self.head == None:
            self.head = new_node
            return

        cur_node = self.head 
        while cur_node.next:
            cur_node = cur_node.next

        cur_node.next = new_node

    def printList(self):
        cur_node = self.head 

        while cur_node:
            print(cur_node.value)
            cur_node = cur_node.next

    def reverseList(self):
        """
            / Before changing next of current, 
            // store next node 
            next = curr->next
            // Now change next of current 
            // This is where actual reversing happens 
            curr->next = prev 
            // Move prev and curr one step forward 
            prev = curr 
            curr = next
        """    
        prev = None 
        curr =self.head 
        nxt = None

        while curr:
            nxt = curr.next
            curr.next = prev

            prev = curr
            curr = nxt

        self.head = prev

   


llist = LinkedList()
llist.append(1)
llist.append(2)
llist.append(4)
llist.append(2)
llist.append(6)
llist.append(5)
# llist.reverseList()
llist.printList()